### PR TITLE
Implements np.unique

### LINKED ISSUE #70: Unimplemented NumPy core functions
Remaining functions to be implemented:

- [x] np.argpartition (tricky because XLA does not provide any partial sort functionality) 
- [x] np.partition (tricky because XLA does not provide any partial sort functionality)

The list above was made by inspecting `jnp._NOT_IMPLEMENTED` and excluding deprecated functions (such as `np.alen`, `np.ipmt`, etc.), functions not relevant to JAX (such as `np.setbufsize`, `np.ascontiguousarray`, etc), and functions that modify buffers in-place (`np.put`, `np.place`, etc.):

Bugs for high-level categories:
 - [x] np.linalg: https://github.com/google/jax/issues/1999
 - [x] np.fft: https://github.com/google/jax/issues/1877
 - [x] searching: https://github.com/google/jax/issues/2080
 - [x] sorting: https://github.com/google/jax/issues/2079
 - [x] sets: https://github.com/google/jax/issues/2078
 - [x] NaN functions: https://github.com/google/jax/issues/2077
 - [x] Index tricks: https://github.com/google/jax/issues/5850

### PR DESCRIPTION
**Pull request by rhitayu98**
Fixes #70 (np.unique)

### CODE REVIEW

**Comment by rhitayu98** [comment]
@shoyer @hawkinsp Kindly review the PR at your convenience and please suggest the changes that I can make.

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -1664,6 +1664,84 @@ def identity(n, dtype=None):
...
+			imask = onp.cumsum(mask) - 1
+			inv_idx = onp.empty(mask.shape, dtype=onp.intp)
+			inv_idx[perm] = imask
+			ret += (inv_idx,)
+		if return_counts:
+			idx = onp.concatenate(onp.nonzero(mask) + ([mask.size],))
+			ret += (onp.diff(idx),)
+		return (ret)
+
+	ar = asanyarray(ar)
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: Unfortunately it's not a good idea to cast to a classic NumPy array (here, you call `onp.array` on the input, which yields a NumPy).

e.g.:
```
In [26]: jax.numpy.ones((3, 4)) + 2                                                                       
Out[26]: 
DeviceArray([[3., 3., 3., 3.],
             [3., 3., 3., 3.],
             [3., 3., 3., 3.]], dtype=float32)

In [27]: asanyarray(jax.numpy.ones((3, 4)) + 2)                                                           
Out[27]: 
array([[3., 3., 3., 3.],
       [3., 3., 3., 3.],
       [3., 3., 3., 3.]], dtype=float32)
```
Note the output is a classic NumPy array.

The problem with doing this is that JAX's tracer doesn't know understand how to trace through classic NumPy ops, only the `lax` primitives and operations built on top of them. So, for example, you wouldn't be able to batch this operation with `vmap` or compile it with `jit`. So you need to express this operation in terms of the operations in `jax.lax` (or other `jax.numpy` operations).

I also note that if we *did* want to implement this operation by casting to a classic NumPy array, then there's a much simpler version...
```
unique = onp.unique
```

**Review by hawkinsp — COMMENTED** [review]
Thanks for the PR. We appreciate it! Unfortunately, I don't think we can accept it as is, assuming I've understood it correctly.

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (rhitayu98) [reply jax/numpy/lax_numpy.py]: Hey th

Comment (rhitayu98) [reply jax/numpy/lax_numpy.py]: Hey there!  In functions like _get_min_identity(dtype) the return type is onp.array therefore i used in the above function. Would it be viable to just typecast the return type in a normal list or use lax.full() with appropriate parameters .

**Comment by rhitayu98** [comment]
Sir, I am a bit unsure if I have understood you correctly, regarding the casting of Numpy arrays. I wanted to be certain if I am heading in the right direction, regarding the code. Here is what I understood.

```
file = numpy.array(trial)
file_new = list(trial)
```

If I return the ___file_new___ in the output, would the Jax tracer be able to iterate through the list and produce the right result or will the Jax tracer not be able to trace the initial Numpy.array, i.e., **_file_**.
Is this the right approach or have I not grasped your point correctly? If not, would you mind being a bit more elaborate.
Thank You

**Comment by hawkinsp** [comment]
JAX can only trace through functions implemented in terms of its own primitives. So, for example, we wouldn't be able to trace through `numpy.array`.

That doesn't mean you can't *use* classic NumPy as part of a JAX program, it just means that if you do so you won't be able to use JAX transformations over that code.

Hope that helps!

**Pull request closed without merging**